dev/integration_tests/windows_startup_test/lib/main.dart

```diff
@@ -70,25 +89,45 @@ void main() async {
       throw 'Window should be hidden at startup';
     }
 
-    bool firstFrame = true;
-    ui.PlatformDispatcher.instance.onBeginFrame = (Duration duration) async {
-      if (await isWindowVisible()) {
-        if (firstFrame) {
-          throw 'Window should be hidden on first frame';
-        }
-
-        if (!visibilityCompleter.isCompleted) {
-          visibilityCompleter.complete('success');
-        }
+    const int totalTestFrames = 10;
+    int frameCount = 0;
+    ui.PlatformDispatcher.instance.onBeginFrame = (Duration duration) {
+      // Our goal is to verify that it's `drawHelloWorld` that makes the window
+      // appear, not anything else. This requires checking the visibility right
+      // before drawing, but since `isWindowVisible` has to be async, and
+      // `FlutterView.render` (in `drawHelloWorld`) forbids async before it,
+      // this can not be done during a single onBeginFrame. However, we can
+      // verify in separate frames to indirectly prove it, by ensuring that
+      // no other mechanism can affect isWindowVisible in the first frame at all.
+      frameCount += 1;
+      switch (frameCount) {
+        // The 1st frame: render nothing, just verify that the window is hidden.
+        case 1:
+          isWindowVisible().then((bool visible) {
+            _expectVisible(visible, false, visibilityCompleter, frameCount);
+            ui.PlatformDispatcher.instance.scheduleFrame();
+          });
+        // The 2nd frame: render, which makes the window appear.
+        case 2:
+          drawHelloWorld(view);
+          ui.PlatformDispatcher.instance.scheduleFrame();
+        // The 3nd frame: render nothing, just wait for the window to appear.
+        case 3:
+          _waitUntilWindowVisible().then((_) {
+            ui.PlatformDispatcher.instance.scheduleFrame();
+          });
+        // Others: Render and verify the window is shown.
+        default:
+          isWindowVisible().then((bool visible) {
+            _expectVisible(visible, true, visibilityCompleter, frameCount);
+            ui.PlatformDispatcher.instance.scheduleFrame();
+            if (frameCount == totalTestFrames && !visibilityCompleter.isCompleted) {
```

Comment:
I wouldn't tie this to a number of test frames. There's no guarantee the window will be visible by the tenth frame, this might cause flakes if the platform thread is exceptionally slow for whatever reason.